lib/internal/vm/Module.js

```diff
@@ -0,0 +1,150 @@
+'use strict';
+
+const { isContext } = process.binding('contextify');
+const errors = require('internal/errors');
+const util = require('util');
+
+const {
+  ModuleWrap,
+  kUninstantiated,
+  kInstantiating,
+  kInstantiated,
+  kEvaluating,
+  kEvaluated,
+  kErrored,
+} = internalBinding('module_wrap');
+
+const StatusMap = {
+  [kUninstantiated]: 'uninstantiated',
+  [kInstantiating]: 'instantiating',
+  [kInstantiated]: 'instantiated',
+  [kEvaluating]: 'evaluating',
+  [kEvaluated]: 'evaluated',
+  [kErrored]: 'errored',
+};
+
+let globalModuleId = 0;
+const kWrap = Symbol('wrap');
+const kContextModuleCount = Symbol('contextModuleCount');
+const kDependencyCache = Symbol('dependencyCache');
+
+class Module {
+  constructor(src, options = {}) {
+    let context;
+    if (options.context !== undefined && isContext(options.context.sandbox)) {
+      context = options.context.sandbox;
+      if (options.context[kContextModuleCount] === undefined)
+        throw new errors.TypeError('ERR_CONTEXT_NOT_MODULE_CONTEXT');
+    }
+
+    const url = options.url ||
+      `vm:module(${context !== undefined ?
+        options.context[kContextModuleCount]++ :
+        globalModuleId++})`;
+
+    const wrap = new ModuleWrap(src, url, {
+      context,
+      lineOffset: options.lineOffset,
+      columnOffset: options.columnOffset,
+    });
+
+    Object.defineProperties(this, {
+      [kWrap]: { value: wrap },
+      context: { value: options.context },
+    });
+  }
+
+  get status() {
+    return StatusMap[this[kWrap].getStatus()];
+  }
+
+  get url() {
+    return this[kWrap].url;
+  }
+
+  get namespace() {
+    if (this[kWrap].getStatus() !== kEvaluated)
+      throw new errors.Error('ERR_VM_MODULE_STATUS', 'must be evaluated');
+    return this[kWrap].namespace();
+  }
+
+  get dependencySpecifiers() {
+    if (this[kDependencyCache] !== undefined)
+      return this[kDependencyCache];
+
+    return this[kDependencyCache] = this[kWrap].getDependencySpecifiers();
+  }
+
+  async link(fn) {
+    const wrap = this[kWrap];
+    if (wrap.isLinked())
+      throw new errors.Error('ERR_VM_MODULE_ALREADY_LINKED');
+    if (wrap.getStatus() !== kUninstantiated)
+      throw new errors.Error('ERR_VM_MODULE_STATUS', 'must be uninstantiated');
+    const promises = [];
+    wrap.link(async (specifier) => {
+      const p = fn(specifier);
+      promises.push(p);
+      const m = await p;
+      if (!(m instanceof Module))
+        throw new errors.Error('ERR_VM_MODULE_NOT_MODULE');
+      return m[kWrap];
+    });
+    await Promise.all(promises);
+  }
+
+  instantiate() {
+    const wrap = this[kWrap];
+    if (wrap.getStatus() !== kUninstantiated)
```

Comment:
The spec allows `"instantiating"`, `"instantiated"`, and `"evaluated"` too as no-ops in those situations. I'm not sure why, but it seems best to keep that compatibility.

Reply:
The spec does this so that the [InnerModuleInstantiation](https://tc39.github.io/ecma262/#sec-innermoduleinstantiation) can be called on things that have already been evaluated/instantiation has started.